healthchecker/healthchecker.go

```diff
@@ -1,57 +1,60 @@
+// healthchecker package provides a funtionality to activelly monitor any target (e.g. file, website).
 package healthchecker
 
 import (
 	"time"
 )
 
-type HealthChecker interface {
-	IsHealthy() bool
-	GetInterval() time.Duration
-	RunSingleCheck()
-}
-
 type SingleChecker interface {
 	SingleCheck() bool
 }
 
 type HealthCheck struct {
+	// Unique check name.
 	Name            string
-	FailedCnt       int
+	// Thresholds define the level of healthy/unhealthy.
 	FailedThreshold int
-	PassedCnt       int
 	PassedThreshold int
-	Healthy         bool
+	// Data polling interval.
 	Interval        time.Duration
+	// A target to monitor.
 	S               SingleChecker
-}
 
-func NewHealthCheck(name string, failedThreshold, passedThreshold int, interval time.Duration, sc SingleChecker) HealthChecker {
-	return &HealthCheck{name, 0, failedThreshold, 0, passedThreshold, false, interval, sc}
-}
+	failedCount     int
+	passedCount     int
+	healthy         bool
 
-func (hc HealthCheck) GetInterval() time.Duration {
-	return hc.Interval
-}
 
-func (hc HealthCheck) String() string {
-	return hc.Name
 }
 
-func (hc HealthCheck) IsHealthy() bool { return hc.Healthy }
+func (hc HealthCheck) IsHealthy() bool { return hc.healthy }
+
+func (hc HealthCheck) Run(ctx context.Context) {
+	go func() {
+		for {
+			select {
+			case <-time.After(hc.Interval):
+				hc.runSingleCheck()
+				fmt.Printf("Name: %q\tHealth: %v\n", hc.Name, hc.IsHealthy())
+			case <-ctx.Done():
+				fmt.Printf("Name: %q\tContext terminated", hc.Name)
```

Comment:
shouldn't this return?

Reply:
That's right. I would include an error return in the function signature & change the line to be:
case <- ctx.Done:
  [...]whatever print
  return fmt.Errorf("Context terminated") // or ctx.Err()